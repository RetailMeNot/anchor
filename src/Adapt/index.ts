export { Adapt } from './Adapt.component';
