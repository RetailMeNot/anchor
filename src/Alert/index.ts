export { Alert } from './Alert.component';
