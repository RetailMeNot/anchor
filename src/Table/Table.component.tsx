// VENDOR
import * as React from 'react';
import classNames from 'classnames';
import styled, { css } from '@xstyled/styled-components';
import {
    variant as createVariant,
    margin as marginStyles,
    MarginProps,
    width as widthStyles,
    WidthProps,
    height as heightStyles,
    HeightProps,
    maxWidth as maxWidthStyles,
    MaxWidthProps,
    minWidth as minWidthStyles,
    MinWidthProps,
    padding as paddingStyles,
    PaddingProps,
    textAlign as textAlignStyles,
    TextAlignProps,
    fontSize as fontSizeStyles,
    FontSizeProps,
    fontWeight as fontWeightStyles,
    FontWeightProps,
    color as colorStyles,
} from '@xstyled/system';

interface SizeTheme {
    [K: string]: {
        cell?: any;
        row?: any;
        header?: any;

        td?: any;
        th?: any;
    };
}

const TableSizes: SizeTheme = {
    sm: {
        th: { padding: '0.25rem 0.75rem' },
        td: { padding: '0.25rem 0.75rem' },
    },
    md: {
        th: { padding: '0.5rem 1.25rem 0.25rem' },
        td: { padding: '0.5rem 1.25rem' },
    },
};

export const sizeVariantStyles = createVariant({
    key: 'table.sizes',
    prop: 'size',
    default: 'md',
    variants: TableSizes,
});

// Table
// ----------------------------------------------------------------------------------

interface TableProps
    extends React.HTMLAttributes<HTMLTableElement>,
        MarginProps,
        WidthProps,
        MaxWidthProps,
        MinWidthProps {
    borderRadius?: string;
    size?: string;
}

const StyledTable = styled('table')<TableProps>`
    border: base;
    border-spacing: 0;

    overflow: hidden;

    color: text.light;
    font-weight: 500;

    ${sizeVariantStyles}
    ${marginStyles}
    ${widthStyles}
    ${maxWidthStyles}
    ${minWidthStyles}
    ${({ borderRadius = 'base' }) =>
        css({
            borderRadius,
        })}

    tbody tr:nth-child(odd) {
        background-color: background.base;
    }
    tr {
        border: dark;
    }

    th {
        white-space: nowrap;
    }
`;

StyledTable.defaultProps = { borderRadius: 'base' };

export const Table = ({
    className,
    children,
    ...props
}: TableProps): React.ReactElement<TableProps> => (
    <StyledTable className={classNames('anchor-table', className)} {...props}>
        {children}
    </StyledTable>
);

// Table.Row
// ----------------------------------------------------------------------------------

interface TableRowProps extends React.HTMLAttributes<HTMLTableRowElement> {}

export const TableRow = ({
    className,
    children,
    ...props
}: TableRowProps): React.ReactElement<TableRowProps> => (
    <tr className={classNames('anchor-table-row', className)} {...props}>
        {children}
    </tr>
);

Table.Row = TableRow;

// Table.Body
// ----------------------------------------------------------------------------------

interface TableBodyProps
    extends React.HTMLAttributes<HTMLTableSectionElement> {}

export const TableBody = ({
    className,
    children,
    ...props
}: TableBodyProps): React.ReactElement<TableBodyProps> => (
    <tbody className={classNames('anchor-table-body', className)} {...props}>
        {children}
    </tbody>
);

Table.Body = TableBody;

// Table.Head
// ----------------------------------------------------------------------------------

interface TableHeadProps
    extends React.HTMLAttributes<HTMLTableSectionElement> {}

export const TableHead = ({
    className,
    children,
    ...props
}: TableHeadProps): React.ReactElement<TableHeadProps> => (
    <thead className={classNames('anchor-table-head', className)} {...props}>
        {children}
    </thead>
);

Table.Head = TableHead;

// Table.Footer
// ----------------------------------------------------------------------------------

interface TableFooterProps
    extends React.HTMLAttributes<HTMLTableSectionElement> {}

export const TableFooter = ({
    className,
    children,
    ...props
}: TableFooterProps): React.ReactElement<TableFooterProps> => (
    <tfoot className={classNames('anchor-table-footer', className)} {...props}>
        {children}
    </tfoot>
);

Table.Footer = TableFooter;

// Table.Cell
// ----------------------------------------------------------------------------------

interface TableCellProps
    extends React.HTMLAttributes<HTMLTableCellElement>,
        FontSizeProps,
        FontWeightProps,
        TextAlignProps,
        WidthProps,
        MaxWidthProps,
        MinWidthProps,
        PaddingProps,
        HeightProps {
    align?: 'center' | 'left' | 'right';
}

const StyledTd = styled('td')<TableCellProps>`
    ${heightStyles}
    ${widthStyles}
    ${maxWidthStyles}
    ${minWidthStyles}
    ${paddingStyles}
    ${fontSizeStyles}
    ${fontWeightStyles}
    ${colorStyles}
    ${({ align }) => textAlignStyles({ textAlign: align })}
`;

export const TableCell = ({
    className,
    children,
    ...props
}: TableCellProps): React.ReactElement<TableCellProps> => (
    <StyledTd className={classNames('anchor-table-cell', className)} {...props}>
        {children}
    </StyledTd>
);

Table.Cell = TableCell;

// Table.HeaderCell
// ----------------------------------------------------------------------------------

interface TableHeaderCellProps
    extends React.HTMLAttributes<HTMLTableHeaderCellElement>,
        FontSizeProps,
        FontWeightProps,
        TextAlignProps,
        WidthProps,
        MaxWidthProps,
        MinWidthProps,
        PaddingProps,
        HeightProps {
    align?: 'center' | 'left' | 'right';
    scope?: string;
}

const StyledTh = styled('th')<TableHeaderCellProps>`
    ${heightStyles}
    ${widthStyles}
    ${maxWidthStyles}
    ${minWidthStyles}
    ${paddingStyles}
    ${fontSizeStyles}
    ${fontWeightStyles}
    ${colorStyles}
    ${({ align }) => textAlignStyles({ textAlign: align })}
`;

export const TableHeaderCell = ({
    className,
    children,
    ...props
}: TableHeaderCellProps): React.ReactElement<TableHeaderCellProps> => (
    <StyledTh
        className={classNames('anchor-table-header-cell', className)}
        {...props}
    >
        {children}
    </StyledTh>
);

Table.HeaderCell = TableHeaderCell;
