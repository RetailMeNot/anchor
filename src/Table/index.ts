export { Table } from './Table.component';
