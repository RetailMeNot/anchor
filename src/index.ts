export { Avatar } from './Avatar';
export { AutoComplete } from './AutoComplete';
export { Button } from './Button/Button.component';
export { Badge } from './Badge';
export { DropDown } from './DropDown';
export { Container, Flex, Box } from './Grid';
export { Icon } from './Icon';
export { Page, Footer, Header } from './Layout';
export { List, Item as ListItem } from './List';
export { Menu, Item as MenuItem } from './Menu';
export { colors, fonts, sizes, variables, GlobalCSS, NormalizeCSS } from './theme';
