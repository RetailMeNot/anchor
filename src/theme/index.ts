export { default as colors } from './colors.theme';
export { default as fonts } from './fonts.theme';
export { default as sizes } from './sizes.theme';
export { default as variables } from './variables.theme';

export { NormalizeCSS, GlobalCSS } from './GlobalStyles';
