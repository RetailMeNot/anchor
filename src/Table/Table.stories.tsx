// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
import { select, text } from '@storybook/addon-knobs';
// VENDOR
import styled, { ThemeProvider } from '@xstyled/styled-components';
// ANCHOR
import { RootTheme } from '../../src/theme';
import { Table } from './Table.component';
import { Typography } from '../Typography';

const { Cell, Row, Head, Body, HeaderCell } = Table;

const StyledStory = styled('div')`
    padding: 2rem;
    color: text.base;
    font-family: base;
`;

const createData = (name: string, color: string, rating: string) => ({
    name,
    color,
    rating,
});
const data = [
    createData('Pear', 'Green', '4/5'),
    createData('Apple', 'Red', '2/5'),
    createData('Banana', 'Yellow', '4/5'),
    createData('Grapefruit', 'Orange', '3.5/5'),
    createData('Apricot', 'Orange', '1.5/5'),
];

storiesOf('Components/Table', module).add('Default', () => (
    <ThemeProvider theme={RootTheme}>
        <StyledStory>
            <Typography
                tag="h3"
                scale={24}
                weight={500}
                mb="0.5rem"
                color="text.light"
            >
                Fruit Table
            </Typography>
            <Table
                width={text('width', '33.75rem')}
                size={select<'md' | 'sm'>('size', ['md', 'sm'], 'md')}
                background={text('background', '') || undefined}
                altBackground={text('altBackground', '') || undefined}
            >
                <Head>
                    <Row>
                        <HeaderCell scope="col" />
                        <HeaderCell scope="col">
                            <Typography scale={28} weight={600} tag="p">
                                $4
                            </Typography>
                            <Typography
                                scale={14}
                                color="text.light"
                                weight="normal"
                            >
                                (30 Day Qty)
                            </Typography>
                        </HeaderCell>
                        <HeaderCell scope="col">
                            <Typography scale={28} weight={600} tag="p">
                                $10
                            </Typography>
                            <Typography
                                scale={14}
                                color="text.light"
                                weight="normal"
                            >
                                (90 Day Qty)
                            </Typography>
                        </HeaderCell>
                    </Row>
                </Head>
                <Body>
                    {data.map(({ name, color, rating }) => (
                        <Row key={name}>
                            <HeaderCell
                                scope="row"
                                align="left"
                                color="#91609c"
                                fontWeight={600}
                            >
                                {name}
                            </HeaderCell>
                            <Cell
                                align="center"
                                width={{ xs: '1px', md: '200px' }}
                            >
                                {color}
                            </Cell>
                            <Cell
                                align="center"
                                width={{ xs: '1px', md: '200px' }}
                            >
                                {rating}
                            </Cell>
                        </Row>
                    ))}
                </Body>
            </Table>
        </StyledStory>
    </ThemeProvider>
));
