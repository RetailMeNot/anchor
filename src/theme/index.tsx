// Theme Files
import { typography } from './typography.theme';
import { ColorsTheme } from './colors.theme';
import { ThemeProvider as XstyledThemeProvider } from '@xstyled/styled-components';
import styled from '@xstyled/styled-components';
import { space as spaceStyles, SpaceProps } from '@xstyled/system';
import { BUTTON_KEY, BUTTON_THEME } from '../Button';
import { INPUT_KEY, INPUT_THEME } from '../Form/Input/utils';
import React from 'react';
import {Button} from '../Button';
import classNames from 'classnames';

export const RootTheme = {
    typography,
    radii: { none: '0', base: '4px', modal: '8px', circular: '1000px' },
    [BUTTON_KEY]: BUTTON_THEME,
    [INPUT_KEY]: INPUT_THEME,
    colors: ColorsTheme,
    fonts: {
        base:
            'Avenir Next, Segoe UI, Helvetica Neue, Helvetica, Roboto, sans-serif',
        mono: `'SFMono-Regular', Consolas, Liberation Mono, Menlo, Courier, monospace`,
    },
    // These are all minimums
    breakpoints: {
        xs: 0,
        sm: 576,
        md: 768,
        lg: 992,
        xl: 1200,
        xxl: 1600,
    },
    // todo: we should likely remove the color.borders colors and just use this
    // or find a way to reference the colors.border colors here
    // e.g. `thin solid borders.base`
    borders: {
        light: `thin solid ${ColorsTheme.borders.light}`,
        base: `thin solid ${ColorsTheme.borders.base}`,
        dark: `thin solid ${ColorsTheme.borders.dark}`,
    },

    skeleton: {
        colorStart: '#E7E7E7',
        colorEnd: '#D3D3D3',
    },
};

// Exporting out @xstyled's ThemeProvider. Only way I could get it to work w/o issues.
// export class ThemeProvider2 extends XstyledThemeProvider {}

interface IterableObject {
    [s: number]: number | string | boolean | IterableObject;
  }

interface DebugProps {
    property: number | string | boolean | IterableObject;
    propertyName: string;
    rootProperty?: boolean;
}

interface StyledDebugProps {
    indent?: number;
}

const StyledDebug = styled('div')<StyledDebugProps>`
    font-family: mono;
    margin-top: 0.5rem;
    color: white;

    .category .category h2 {
        background-color: #ec8947;
    }

    .category .category .box {
        border-color: #ec8947;
    }

    .category .category .category h2 {
        background-color: #b84343;
        color: white;
    }

    .category .category .category .box {
        border-color: #b84343;
    }

    .category .category .category .category h2 {
        background-color: #91cdd2;
        color: black;
    }

    .category .category .category .category .box {
        border-color: #91cdd2;
    }

    .category .category .category .category .category h2 {
        background-color: #567224;
        color: white;
    }

    .category .category .category .category .category .box {
        border-color: #567224;
    }
`;

const StyledH2 = styled('h2')`
    ${spaceStyles}
    background-color: #ffea92;
    color: black;
    margin-bottom: 0;
    margin-top: 0;
    padding: 0.25rem 0.5rem 0.25rem 1rem;
    font-size: 1rem;
    cursor: pointer;
    border-radius: 0 1rem 1rem 0;
`;

interface BoxProps extends SpaceProps {
    show?: boolean;
}

const Box = styled('div')<BoxProps>`
    ${spaceStyles}
    width: 20rem;
    border-left: solid #ffea92 2px;
    margin: 0.5rem;
    ${({show}) => `display: ${show ? 'block' : 'none'}`}
`;

const Category = styled('div')<SpaceProps>`
    ${spaceStyles}
    width: 20rem;
`;


const ColorBox = styled('div')<any>`
    width: 1rem;
    height: 1rem;
    display: inline-block;
    vertical-align: middle;
    margin-left: 0.5rem;
    background-color: ${({backgroundColor}) => backgroundColor};
    border: solid white 1px;
`;

const Debug = ({property, propertyName}: DebugProps) => {
    const [show, setShow] = React.useState(false);

    return(
        <StyledDebug>
            {(property || property === 0)
                ? (
                    typeof property === 'number' ||
                    typeof property === 'string' ||
                    typeof property === 'boolean' ||
                    typeof property === 'function'
                        ? (
                                <>
                                    <span style={{fontWeight: 'bold'}}>{propertyName}: </span>
                                    <span>
                                        {typeof property === 'function'
                                            ? 'function'
                                            : /^#[a-zA-Z0-9]{4,7}/.test(property.toString())
                                                ? (<>
                                                    <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                                    <ColorBox backgroundColor={property} />
                                                    </>)
                                                : <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                        }
                                    </span>
                                </>
                            )
                        :
                            <Category pl={0} className="category">
                                <StyledH2 onClick={() => setShow(!show)}>{propertyName}</StyledH2>
                                <Box pl={2} className="box" show={show}>
                                    {Object.values(property).map((_property, index) => (
                                        <Debug
                                            key={index}
                                            property={_property}
                                            propertyName={Object.getOwnPropertyNames(property)[index]}
                                        />
                                    ))}
                                </Box>
                            </Category>
                    )
                : (
                    <>
                       Empty.
                    </>
                )
            }
        </StyledDebug>
    );
};

const DebugWrapper = styled('div')`
    background-color: rgba(0,0,0,.75);
    border-left-color: transparent;
    position: fixed;
    left: 0;
    top: 0;
    height: 100%;
    overflow: hidden;
    width: 30rem;
    padding: 1rem 2rem 1rem 0;

    &.closed {
        height:2.4rem;
        width: auto;
        padding-right: 1rem;

        .debug-scroll {
            display: none;
        }
    }
`;

const DebugScroll = styled('div')`
    height: 100%;
    overflow-y: auto;
    width: 100%;
`;

interface ThemeProviderProps {
    children?: any;
    debug?: boolean;
    theme?: any;
}

export const ThemeProvider = ({ children, debug, theme }: ThemeProviderProps) => {
    const isOpen = localStorage.themeProviderDebug === 'true';
    const [isOpenState, setIsOpenState] = React.useState(isOpen);

    return (
        <XstyledThemeProvider theme={theme}>
            {debug &&
                <DebugWrapper className={classNames(!isOpenState ? 'closed' : null)}>
                    <Button
                        onClick={() => {
                            localStorage.themeProviderDebug = !isOpenState;
                            setIsOpenState(!isOpenState);
                        }}
                        size="sm"
                    >
                        ThemeProvider Debug
                    </Button>

                    <DebugScroll className="debug-scroll">
                        <Debug property={theme} propertyName="Theme" />
                    </DebugScroll>
                </DebugWrapper>
            }
            {children}
        </XstyledThemeProvider>
    );
};
