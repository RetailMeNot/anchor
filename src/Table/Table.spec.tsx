// VENDOR
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ThemeProvider } from '@xstyled/styled-components';
import { shallow, mount } from 'enzyme';
// ANCHOR
import { RootTheme } from '../../src/theme';
import { Table } from './Table.component';
const { Head, Body, Cell, Footer, Row, HeaderCell } = Table;

describe('Component: Table', () => {
    const table = (
        <ThemeProvider theme={RootTheme}>
            <Table />
        </ThemeProvider>
    );

    const wrapper = mount(table);
    const component = shallow(table);

    expect(table).toBeDefined();
    expect(wrapper).toBeDefined();
    expect(component).toBeDefined();

    it('should match snapshot.', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <Table />
            </ThemeProvider>
        );

        const tree = renderer.create(subject).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render with all the bells and whistles', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <Table>
                    <Head>
                        <Row>
                            <HeaderCell>Fruit</HeaderCell>
                            <HeaderCell>Color</HeaderCell>
                            <HeaderCell>Calories</HeaderCell>
                        </Row>
                    </Head>
                    <Body>
                        <Row>
                            <Cell>Apple</Cell>
                            <Cell>Red</Cell>
                            <Cell>95</Cell>
                        </Row>
                        <Row>
                            <Cell>Banana</Cell>
                            <Cell>Yellow</Cell>
                            <Cell>105</Cell>
                        </Row>
                        <Row>
                            <Cell>Cranberry</Cell>
                            <Cell>Red</Cell>
                            <Cell>2</Cell>
                        </Row>
                    </Body>
                    <Footer>
                        <Row>
                            <Cell>People use tfoots?</Cell>
                        </Row>
                    </Footer>
                </Table>
            </ThemeProvider>
        );

        const tree = renderer.create(subject).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
