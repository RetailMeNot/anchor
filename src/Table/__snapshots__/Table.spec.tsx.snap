// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Table should be defined. 1`] = `
.c0 {
  border: thin solid #D3D3D3;
  border-spacing: 0;
  overflow: auto;
  color: #595959;
  font-weight: 500;
  border-radius: 4px;
}

.c0 th {
  padding: 0.5rem 1.25rem 0.25rem;
}

.c0 td {
  padding: 0.5rem 1.25rem;
}

.c0 tbody tr:nth-child(odd) {
  background-color: #F1F1F1;
}

.c0 tr {
  border: thin solid #808080;
}

.c0 th {
  white-space: nowrap;
}

<table
  className="c0 anchor-table"
/>
`;

exports[`Component: Table should render with all the bells and whistles 1`] = `
.c0 {
  border: thin solid #D3D3D3;
  border-spacing: 0;
  overflow: auto;
  color: #595959;
  font-weight: 500;
  border-radius: 4px;
}

.c0 th {
  padding: 0.5rem 1.25rem 0.25rem;
}

.c0 td {
  padding: 0.5rem 1.25rem;
}

.c0 tbody tr:nth-child(odd) {
  background-color: #F1F1F1;
}

.c0 tr {
  border: thin solid #808080;
}

.c0 th {
  white-space: nowrap;
}

<table
  className="c0 anchor-table"
>
  <thead
    className="anchor-table-head"
  >
    <tr
      className="anchor-table-row"
    >
      <th
        className="anchor-table-header-cell"
      >
        Fruit
      </th>
      <th
        className="anchor-table-header-cell"
      >
        Color
      </th>
      <th
        className="anchor-table-header-cell"
      >
        Calories
      </th>
    </tr>
  </thead>
  <tbody
    className="anchor-table-body"
  >
    <tr
      className="anchor-table-row"
    >
      <td
        className="anchor-table-cell"
      >
        Apple
      </td>
      <td
        className="anchor-table-cell"
      >
        Red
      </td>
      <td
        className="anchor-table-cell"
      >
        95
      </td>
    </tr>
    <tr
      className="anchor-table-row"
    >
      <td
        className="anchor-table-cell"
      >
        Banana
      </td>
      <td
        className="anchor-table-cell"
      >
        Yellow
      </td>
      <td
        className="anchor-table-cell"
      >
        105
      </td>
    </tr>
    <tr
      className="anchor-table-row"
    >
      <td
        className="anchor-table-cell"
      >
        Cranberry
      </td>
      <td
        className="anchor-table-cell"
      >
        Red
      </td>
      <td
        className="anchor-table-cell"
      >
        2
      </td>
    </tr>
  </tbody>
  <tfoot
    className="anchor-table-footer"
  >
    <tr
      className="anchor-table-row"
    >
      <td
        className="anchor-table-cell"
      >
        People use tfoots?
      </td>
    </tr>
  </tfoot>
</table>
`;
